[7.4.0] Crash during exec log generation

### Description of the bug:


```
-- MODULE.bazel --
-- pkg/BUILD --
DATA = glob(["*.txt"])

sh_binary(
    name = "tool",	
    srcs = ["tool.sh"],
    data = DATA,
    args = ["$(rootpath {})".format(file) for file in DATA],
)
genrule(
    name = "gen1",
    outs = ["out1"],
    cmd = "$(location :tool) > $@",
    tools = [":tool"],
)
genrule(
    name = "gen2",
    outs = ["out2"],
    cmd = """
$(location :tool) > $@
$(location :tool) >> $@
""",
    tools = [":tool"],
)
-- pkg/tool.sh --
#!/bin/bash
cat "$@"
-- pkg/file1.txt --
old
-- pkg/file2.txt --
unchanged
```

results in:

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData0{actionLookupKey=ConfiguredTargetKey{label=//pkg:gen1, config=BuildConfigurationKey[4cf1c944324811e840e4f5049a0d662eb35b2ebf65bdb7fdb6de7d17810fb043]}, actionIndex=0}' (requested by nodes 'TargetCompletionKey{topLevelArtifactContext=com.google.devtools.build.lib.analysis.TopLevelArtifactContext@90904c3b, actionLookupKey=ConfiguredTargetKey{label=//pkg:gen1, config=BuildConfigurationKey[4cf1c944324811e840e4f5049a0d662eb35b2ebf65bdb7fdb6de7d17810fb043]}, willTest=false}')
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException: Cannot invoke "com.google.devtools.build.lib.actions.RunfilesSupplier$RunfilesTree.getExecPath()" because "runfilesTree" is null
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logRunfilesTree(CompactSpawnLogContext.java:561)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.lambda$logInputSet$3(CompactSpawnLogContext.java:409)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logEntry(CompactSpawnLogContext.java:712)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logInputSet(CompactSpawnLogContext.java:385)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.lambda$logInputSet$3(CompactSpawnLogContext.java:394)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logEntry(CompactSpawnLogContext.java:695)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logInputSet(CompactSpawnLogContext.java:385)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logInputs(CompactSpawnLogContext.java:328)
	at com.google.devtools.build.lib.exec.CompactSpawnLogContext.logSpawn(CompactSpawnLogContext.java:229)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:191)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:118)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.execute(SpawnAction.java:263)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1159)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1076)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:573)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:862)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:334)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:172)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
	... 7 more
```

### Which category does this issue belong to?

Performance

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


_No response_

### Which operating system are you running Bazel on?


_No response_

### What is the output of `bazel info release`?


_No response_

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


_No response_

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java
+++ b/src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java
@@ -19,7 +19,6 @@ import static com.google.common.base.Preconditions.checkState;
 import com.github.luben.zstd.ZstdOutputStream;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Iterables;
 import com.google.devtools.build.lib.actions.AbstractAction;
 import com.google.devtools.build.lib.actions.ActionInput;
@@ -65,7 +64,6 @@ import java.util.Collection;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
-import java.util.Set;
 import java.util.SortedMap;
 import java.util.concurrent.ForkJoinPool;
 import javax.annotation.Nullable;
@@ -328,7 +326,6 @@ public class CompactSpawnLogContext extends SpawnLogContext {
     return logInputSet(
         spawn.getInputFiles(),
         spawn.getRunfilesSupplier().getRunfilesTreesForLogging(),
-        spawn.getRunfilesSupplier().getRunfilesTreesForLogging().keySet(),
         additionalInputIds.build(),
         inputMetadataProvider,
         fileSystem,
@@ -348,7 +345,6 @@ public class CompactSpawnLogContext extends SpawnLogContext {
     return logInputSet(
         spawn.getToolFiles(),
         spawn.getRunfilesSupplier().getRunfilesTreesForLogging(),
-        ImmutableSet.of(),
         ImmutableList.of(),
         inputMetadataProvider,
         fileSystem,
@@ -371,14 +367,13 @@ public class CompactSpawnLogContext extends SpawnLogContext {
   private int logInputSet(
       NestedSet<? extends ActionInput> set,
       Map<Artifact, RunfilesTree> runfilesTrees,
-      Set<Artifact> extraMiddleman,
       Collection<Integer> additionalDirectoryIds,
       InputMetadataProvider inputMetadataProvider,
       FileSystem fileSystem,
       boolean shared,
       boolean isTestRunnerSpawn)
       throws IOException, InterruptedException {
-    if (set.isEmpty() && additionalDirectoryIds.isEmpty() && extraMiddleman.isEmpty()) {
+    if (set.isEmpty() && additionalDirectoryIds.isEmpty()) {
       return 0;
     }
 
@@ -393,8 +388,7 @@ public class CompactSpawnLogContext extends SpawnLogContext {
             builder.addTransitiveSetIds(
                 logInputSet(
                     transitive,
-                    ImmutableMap.of(),
-                    ImmutableSet.of(),
+                    runfilesTrees,
                     ImmutableList.of(),
                     inputMetadataProvider,
                     fileSystem,
@@ -402,7 +396,7 @@ public class CompactSpawnLogContext extends SpawnLogContext {
                     /* isTestRunnerSpawn= */ false));
           }
 
-          for (ActionInput input : Iterables.concat(set.getLeaves(), extraMiddleman)) {
+          for (ActionInput input : set.getLeaves()) {
             if (input instanceof Artifact artifact && artifact.isMiddlemanArtifact()) {
               RunfilesTree runfilesTree = runfilesTrees.get(artifact);
               builder.addInputIds(
@@ -569,8 +563,8 @@ public class CompactSpawnLogContext extends SpawnLogContext {
           builder.setInputSetId(
               logInputSet(
                   runfilesTree.getArtifactsAtCanonicalLocationsForLogging(),
+                  // Runfiles trees are never nested.
                   ImmutableMap.of(),
-                  ImmutableSet.of(),
                   ImmutableList.of(),
                   inputMetadataProvider,
                   fileSystem,

--- a/src/test/java/com/google/devtools/build/lib/exec/CompactSpawnLogContextTest.java
+++ b/src/test/java/com/google/devtools/build/lib/exec/CompactSpawnLogContextTest.java
@@ -194,13 +194,13 @@ public final class CompactSpawnLogContextTest extends SpawnLogContextTestBase {
     Spawn firstSpawn =
         defaultSpawnBuilder()
             .withRunfilesSupplier(runfilesSupplier)
-            .withInputs(firstInput)
+            .withInputs(firstInput, runfilesMiddleman)
             .withTool(runfilesMiddleman)
             .build();
     Spawn secondSpawn =
         defaultSpawnBuilder()
             .withRunfilesSupplier(runfilesSupplier)
-            .withInputs(secondInput)
+            .withInputs(secondInput, runfilesMiddleman)
             .withTool(runfilesMiddleman)
             .build();
 

--- a/src/test/java/com/google/devtools/build/lib/exec/SpawnLogContextTestBase.java
+++ b/src/test/java/com/google/devtools/build/lib/exec/SpawnLogContextTestBase.java
@@ -421,7 +421,8 @@ public abstract class SpawnLogContextTestBase {
     RunfilesSupplier runfilesSupplier =
         createRunfilesSupplier(runfilesMiddleman, runfilesRoot, runfilesInput);
 
-    SpawnBuilder spawnBuilder = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier);
+    SpawnBuilder spawnBuilder =
+        defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).withInput(runfilesMiddleman);
     if (inputsMode.isTool()) {
       spawnBuilder.withTool(runfilesMiddleman);
     }
@@ -454,6 +455,73 @@ public abstract class SpawnLogContextTestBase {
   }
 
   @Test
+  public void testRunfilesNestedMiddleman() throws Exception {
+    Artifact runfilesMiddleman = ActionsTestUtil.createArtifact(middlemanDir, "runfiles");
+    Artifact runfilesInput = ActionsTestUtil.createArtifact(rootDir, "data.txt");
+    writeFile(runfilesInput, "abc");
+    Artifact toolFile1 = ActionsTestUtil.createArtifact(rootDir, "tool1");
+    writeFile(toolFile1, "def");
+    Artifact toolFile2 = ActionsTestUtil.createArtifact(rootDir, "tool2");
+    writeFile(toolFile2, "ghi");
+
+    PathFragment runfilesRoot = outputDir.getExecPath().getRelative("foo.runfiles");
+    RunfilesSupplier runfilesSupplier =
+        createRunfilesSupplier(runfilesMiddleman, runfilesRoot, runfilesInput);
+
+    NestedSet<ActionInput> tools =
+        NestedSetBuilder.<ActionInput>stableOrder()
+            .add(toolFile1)
+            .addTransitive(
+                NestedSetBuilder.<ActionInput>stableOrder()
+                    .add(runfilesMiddleman)
+                    .add(toolFile2)
+                    .build())
+            .build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInputs(tools)
+            .withTools(tools)
+            .build();
+
+    SpawnLogContext context = createSpawnLogContext();
+    InputMetadataProvider inputMetadataProvider =
+        createInputMetadataProvider(runfilesSupplier, runfilesInput, toolFile1, toolFile2);
+
+    context.logSpawn(
+        spawn,
+        inputMetadataProvider,
+        createInputMap(runfilesSupplier, inputMetadataProvider, toolFile1, toolFile2),
+        fs,
+        defaultTimeout(),
+        defaultSpawnResult());
+
+    closeAndAssertLog(
+        context,
+        defaultSpawnExecBuilder()
+            .addInputs(
+                File.newBuilder()
+                    .setPath(
+                        PRODUCT_NAME
+                            + "-out/k8-fastbuild/bin/foo.runfiles/"
+                            + WORKSPACE_NAME
+                            + "/data.txt")
+                    .setDigest(getDigest("abc"))
+                    .setIsTool(true))
+            .addInputs(
+                File.newBuilder()
+                    .setPath("tool1")
+                    .setDigest(getDigest("def"))
+                    .setIsTool(true))
+            .addInputs(
+                File.newBuilder()
+                    .setPath("tool2")
+                    .setDigest(getDigest("ghi"))
+                    .setIsTool(true))
+            .build());
+  }
+
+  @Test
   public void testRunfilesDirectoryInput(
       @TestParameter DirContents dirContents, @TestParameter InputsMode inputsMode)
       throws Exception {
@@ -469,7 +537,8 @@ public abstract class SpawnLogContextTestBase {
     RunfilesSupplier runfilesSupplier =
         createRunfilesSupplier(runfilesMiddleman, runfilesRoot, runfilesInput);
 
-    SpawnBuilder spawnBuilder = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier);
+    SpawnBuilder spawnBuilder =
+        defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).withInput(runfilesMiddleman);
     if (inputsMode.isTool()) {
       spawnBuilder.withTool(runfilesMiddleman);
     }
@@ -537,7 +606,8 @@ public abstract class SpawnLogContextTestBase {
             externalGenArtifact,
             externalSourceArtifact);
 
-    SpawnBuilder spawnBuilder = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier);
+    SpawnBuilder spawnBuilder =
+        defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).withInput(runfilesMiddleman);
     if (inputsMode.isTool()) {
       spawnBuilder.withTool(runfilesMiddleman);
     }
@@ -687,7 +757,11 @@ public abstract class SpawnLogContextTestBase {
             externalSourceArtifact,
             externalGenArtifact);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -833,7 +907,11 @@ public abstract class SpawnLogContextTestBase {
             externalSourceArtifact,
             externalGenArtifact);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -967,7 +1045,11 @@ public abstract class SpawnLogContextTestBase {
             externalSourceArtifact,
             externalGenArtifact);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -1074,7 +1156,11 @@ public abstract class SpawnLogContextTestBase {
             externalSourceArtifact,
             externalGenArtifact);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -1168,7 +1254,11 @@ public abstract class SpawnLogContextTestBase {
             /* legacyExternalRunfiles= */ false,
             sourceArtifact);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -1217,7 +1307,11 @@ public abstract class SpawnLogContextTestBase {
             /* legacyExternalRunfiles= */ false,
             NestedSetBuilder.wrap(Order.STABLE_ORDER, artifacts));
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -1295,7 +1389,11 @@ public abstract class SpawnLogContextTestBase {
             /* legacyExternalRunfiles= */ false,
             artifacts);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =
@@ -1376,7 +1474,8 @@ public abstract class SpawnLogContextTestBase {
                 : ImmutableMap.of(),
             /* legacyExternalRunfiles= */ false);
 
-    SpawnBuilder spawnBuilder = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier);
+    SpawnBuilder spawnBuilder =
+        defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).withInput(runfilesMiddleman);
     if (inputsMode.isTool()) {
       spawnBuilder.withTool(runfilesMiddleman);
     }
@@ -1455,7 +1554,11 @@ public abstract class SpawnLogContextTestBase {
                 : ImmutableMap.of(),
             /* legacyExternalRunfiles= */ false);
 
-    Spawn spawn = defaultSpawnBuilder().withRunfilesSupplier(runfilesSupplier).build();
+    Spawn spawn =
+        defaultSpawnBuilder()
+            .withRunfilesSupplier(runfilesSupplier)
+            .withInput(runfilesMiddleman)
+            .build();
 
     SpawnLogContext context = createSpawnLogContext();
     InputMetadataProvider inputMetadataProvider =

